src/main/java/net/neoforged/neoforge/fluids/crafting/UnitAmountFluidIngredient.java

```diff
@@ -0,0 +1,267 @@
+
+    public static final StreamCodec<RegistryFriendlyByteBuf, UnitAmountFluidIngredient> STREAM_CODEC = StreamCodec.composite(
+            FluidIngredient.STREAM_CODEC,
+            UnitAmountFluidIngredient::ingredient,
+            ByteBufCodecs.<ByteBuf, FluidUnit, Integer, Map<FluidUnit, Integer>>map(HashMap::new, FluidUnit.STREAM_CODEC, ByteBufCodecs.INT).apply(ByteBufCodecs.list()),
+            UnitAmountFluidIngredient::amounts,
+            UnitAmountFluidIngredient::new);
+
+    /**
+     * Helper method to create a simple unit-based ingredient that has a single unit quantity.
+     */
+    public static UnitAmountFluidIngredient of(FluidIngredient fluid, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(fluid, List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(Fluid fluid, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.of(fluid), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(TagKey<Fluid> tag, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.tag(tag), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Factory method for creating any ingredient.
+     */
+    @SafeVarargs
+    public static UnitAmountFluidIngredient of(FluidIngredient ingredient, Map<FluidUnit, Integer>... amounts) {
+        if (amounts.length == 0)
+            return new UnitAmountFluidIngredient(ingredient, DEFAULT_AMOUNT);
+        else
+            return new UnitAmountFluidIngredient(ingredient, Arrays.stream(amounts)
+                    .map(UnitAmountFluidIngredient::validateNotEmpty)
+                    .map(DataResult::getOrThrow).toList());
+    }
+
+    private final FluidIngredient ingredient;
+    private final List<Map<FluidUnit, Integer>> amounts;
+
+    @Nullable
+    private FluidStack[] cachedStacks;
+
+    private UnitAmountFluidIngredient(FluidIngredient ingredient, List<Map<FluidUnit, Integer>> amounts) {
+        this.ingredient = ingredient;
+        this.amounts = amounts;
+    }
+
+    public FluidIngredient ingredient() {
+        return ingredient;
+    }
+
+    public List<Map<FluidUnit, Integer>> amounts() {
+        return amounts;
+    }
+
+    public Optional<Integer> amount(FluidStack stack) {
+        FluidUnitSpec fluidUnits = stack.getFluidType().getUnits();
+        for (Map<FluidUnit, Integer> amount : amounts) {
+            int accumulator = 0;
+            boolean valid = true;
+            for (Map.Entry<FluidUnit, Integer> entry : amount.entrySet()) {
+                Integer value = fluidUnits.getValue(entry.getKey());
+                if (value == null)
+                    valid = false;
```

Comment:
why no break here? we want fluid ingredients to match or fail to match as fast as possible (if its not valid no point looking at the remaining units or continuing to math).

If we wanted to be fancy, we could write this whole loop using labels and cut out some of the extra variables/if statements, e.g.

```java
    public Optional<Integer> amount(FluidStack stack) {
        FluidUnitSpec fluidUnits = stack.getFluidType().getUnits();
        unitOptions:
        for (Map<FluidUnit, Integer> amount : amounts) {
            int accumulator = 0;
            for (Map.Entry<FluidUnit, Integer> entry : amount.entrySet()) {
                Integer value = fluidUnits.getValue(entry.getKey());
                if (value == null)
                    continue unitOptions;
                accumulator += value * entry.getValue();
            }
            return Optional.of(accumulator);
        }
        return Optional.empty();
    }
```